keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,674 @@
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+    - [ ] e2e Tests for all Beta API Operations (endpoints)
+    - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+    - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+    - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The proposal aims at enabling hot plugging of node compute resources. This will help in updating cluster resource capacity by just resizing compute resources of nodes rather than adding new node to a cluster.
+The updated node configurations are to be reflected at the node and cluster levels automatically without the need to reset the kubelet.
+
+This proposal also aims to improve the initialization and reinitialization of resource managers, such as the CPU manager and memory manager, in response to changes in a node's CPU and memory configurations.
+
+## Motivation
+In a typical Kubernetes environment, the cluster resources may need to be altered due to following reasons:
+- Incorrect resource assignment during cluster creation.
+- Increased workload over time, leading to the need for additional resources in the cluster.
+
+To handle these scenarios, we can:
+- Horizontally scale up the cluster by adding compute nodes.
+- Vertically scale up the cluster by increasing node capacity. However, currently, the workaround for capturing node resizing in the cluster involves restarting the Kubelet.
+
+Node resource hot plugging will provide advantages in scenarios such as:
+- Handling resource demand with a limited set of nodes by increasing the capacity of existing nodes instead of creating new nodes.
+- Creating new nodes takes more time compared to increasing the capacity of existing nodes.
+
+### Goals
```

Comment:
what about swap calculations? Will they be updated if node size increased?

Reply:
Yeah, It should be updated as the swap calculation [depends](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go#L459-L473) on node total memory. I will update the KEP to include the same, Thanks.

Reply:
so we will need to update every Pod with the new swap values? What CRI call we will use to do it?

Reply:
Initially we decided to calculate new swap limit for newly created pods post resize, But we are also considering need for recalculating the swap limit for already existing pods, This is inline to a discussion regarding oom_score_adj: https://github.com/kubernetes/enhancements/pull/3955#discussion_r1942333326

Looking forward to hearing your opinion about it.

Also in-case if we want to update the swap limit for existing pods we are considering using UpdateContainerResources CRI method which is used by Inplace VPA.

Reply:
+1 resizing swap. I think the goal of this kep should be: when a node's resources are hotplugged, the container's resources should be as close to what it would be if that was the initial state as possible, and any gaps should be both justified and very documented

Reply:
Sure, I have accounted the points discuss about updating swap and OOMScoreAdj in the KEP.

Reply:
I do not see the update in the goals section. Maybe I am missing something

Reply:
We added a 3rd goal as 
```
* Recalculating and updating the OOMScoreAdj and swap memory limit for existing pods.
```